mono/metadata/appdomain.c

```diff
@@ -337,25 +343,31 @@ mono_context_set_default_context (MonoDomain *domain)
 	HANDLE_FUNCTION_RETURN ();
 }
 
-
 static int
 mono_get_corlib_version (void)
 {
+	HANDLE_FUNCTION_ENTER ();
 	ERROR_DECL (error);
+
 	MonoClass *klass;
 	MonoClassField *field;
-	MonoObject *value;
+	MonoObjectHandle value;
+	int result = - 1;
 
 	klass = mono_class_load_from_name (mono_defaults.corlib, "System", "Environment");
 	mono_class_init (klass);
 	field = mono_class_get_field_from_name (klass, "mono_corlib_version");
 	if (!field)
-		return -1;
+		goto exit;
+
 	if (! (field->type->attrs & FIELD_ATTRIBUTE_STATIC))
-		return -1;
-	value = mono_field_get_value_object_checked (mono_domain_get (), field, NULL, error);
+		goto exit;
+
+	value = mono_static_field_get_value_handle (mono_domain_get (), field, error);
 	mono_error_assert_ok (error);
-	return *(gint32*)((gchar*)value + sizeof (MonoObject));
+	result = *(gint32*)mono_handle_unbox_unsafe (value);
```

Comment:
I don't like using things called "unsafe".

Reply:
Its unsafe in the same sense as `MONO_HANDLE_RAW()` - you have a raw pointer to managed memory that is not pinned.  There is no `mono_handle_unbox` because it doesn't make sense as a general operation.

In this case we want an analog of `mono_value_copy ()` that takes a handle for a source: `mono_value_copy_from_handle (gpointer dest, MonoObjectHandle src, MonoType *src_type)` where `src_type` must be a primitive or a valuetype.